File: php/RestClient.php
<?php
namespace LianaTech;

require_once __DIR__ . '/APIException.php';
require_once __DIR__ . '/RestClientAuthorizationException.php';

/**
 * PHP RestClient for LianaTech RESTful services
 *
 * Example:
 *
 * 	$client = new RestCLient(...);
 *
 */
class RestClient {

	protected $api_key = null;
	protected $api_user = null;
	protected $api_url = null;
	protected $api_version = null;
	protected $api_realm = null;

	public function __construct($api_user, $api_key, $api_url, $api_version, $api_realm) {
		$this->api_user = $api_user;
		$this->api_key = $api_key;
		$this->api_url = $api_url;
		$this->api_version = intval($api_version);
		$this->api_realm = $api_realm;
	}

	public function call($path, $args = array(), $method = 'POST') {
		$args = $method !== 'GET' ? json_encode($args) : '';
		return $this->request($path, $args, $method);
	}

	protected function sign(array $message) {
		return hash_hmac('sha256', implode("\n", $message), $this->api_key);
	}

	protected function request($path, $contents, $method) {
		$md5 = md5($contents);
		$datetime = new \DateTime('now', new \DateTimeZone('Europe/Helsinki'));
		$timestamp = $datetime->format('c');
		$url = $this->api_url . '/api/v'. $this->api_version .'/' . $path;

		$message = array(
			$method,
			$md5,
			'application/json',
			$timestamp,
			$contents,
			'/api/v'. $this->api_version .'/' . $path
		);

		$signature = $this->sign($message);
		$user_id = $this->api_user;

		$ch = curl_init();
		curl_setopt($ch, CURLOPT_URL, $url);
		curl_setopt($ch, CURLOPT_HTTPHEADER, array(
			'Content-Type: application/json',
			"Content-MD5: {$md5}",
			"Date: {$timestamp}",
			"Authorization: {$this->api_realm} {$user_id}:{$signature}"
		));
		curl_setopt($ch, CURLOPT_USERAGENT, 'PHP-LMAPI');
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
		curl_setopt($ch, CURLOPT_TIMEOUT, 10);
		curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
		curl_setopt($ch, CURLOPT_POSTFIELDS, $contents);
		$result = curl_exec($ch);
		$http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
		curl_close($ch);

		$result = json_decode($result, true);

		switch ($http_code) {
			case 401:
				throw new RestClientAuthorizationException;
				break;
			default:
				if ($http_code >= 400) {
					throw new APIException(sprintf(
						'API response with status code %s %s',
						$http_code,
						$result['message'] ? $result['message'] : ''
					));
				}
		}

		if ($result === null) {
			throw new APIException('API did not return a valid json string');
		}

		return $result;
	}

}

File: python/RestClient.py
"""

Liana REST API client

Copyright Liana Technologies Ltd 2018

"""

import json
import hashlib
import hmac
import requests
import time

class APIException(Exception):
    pass

class RestClient:
    def __init__(self, user_id, api_secret, api_url, api_version, api_realm):
        self._response = None
        self._user_id = user_id
        self._api_secret = api_secret
        self._api_url = api_url
        self._api_realm = api_realm
        self._api_version = api_version
        self._content_type = 'application/json'

    def call(self, path, params=[], method='POST'):
        """ Perform API request and return the API result"""
        request_function = getattr(requests, method.lower())
        self._set_request_data(path, params, method)

        self._response = request_function(
            self._api_url + self._full_path,
            headers=self._get_headers(),
            data=self._json_string
        )

        self._response_body = self._response.text;

        if (self._response.status_code >= 400):
            raise APIException(
                'API response with status code ' +str(self._response.status_code)
             )

        try:
            data = json.loads(self._response_body);
        except ValueError: # Python 2.x
            raise APIException('API did not return a valid json string')
        except json.decoder.JSONDecodeError: # Python 3.5+
            raise APIException('API did not return a valid json string')

        return data

    def get_http_response(self):
        """ Returns the raw response object of last performed API request """
        return self._response



    """ INTERNAL METHODS FOLLOW """

    def _get_new_timestamp(self):
        """ Returns a fresh timestamp in proper format """
        return time.strftime('%Y-%m-%dT%H:%M:%S%z')

    def _get_hash(self):
        """ Form and return the parameters hash for the API request """
        md5 = hashlib.md5()
        md5.update(self._json_string.encode('utf-8'))
        return md5.hexdigest()


    def _get_message(self):
        """ Return the message in the format which is used to create signature of the request """
        message = "\n".join([
            self._method,
            self._get_hash(),
            self._content_type,
            self._timestamp,
            self._json_string,
            self._full_path
        ])
        return message.encode('utf-8')

    def _get_signature(self):
        """ Get signature for the API request """
        return hmac.new(
            self._api_secret.encode('utf-8'),
            self._get_message(),
            hashlib.sha256
        ).hexdigest()


    def _get_headers(self):
        """ Get headers for the API HTTP request """
        return {
            'Content-Type': self._content_type,
            'Content-MD5': self._get_hash(),
            'Date': self._timestamp,
            'Authorization': self._api_realm + ' ' + str(self._user_id) + ':' + self._get_signature(),
        }


    def _set_request_data(self, path, params, method):
        """ Set API request data """
        self._full_path = '/api/v' + str(self._api_version) + '/' + path
        self._json_string = json.dumps(params)
        if method == 'GET':
			self._json_string = ''
        self._timestamp = self._get_new_timestamp()
        self._method = method

